Move Pod Lifecycle to Concepts.

steveperry-53 opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes.github.io/2420)
<!-- Reviewable:end -->

steveperry-53 commented:
@chenopis, @pwittrock

chenopis commented on docs/concepts/workloads/pods/pod-lifecycle.md:
```diff
@@ -0,0 +1,281 @@
+
+* Success: The Container passed the diagnostic.
+* Failure: The Container failed the diagnostic.
+* Unknown: The diagnostic failed, so no action should be taken.
```
This explanation seems confusing to me. So, the diagnostic failed, but for unknown reasons, so no action should be taken? Does that mean this is an acceptable result and should be ignored?

steveperry-53 replied:
@pwittrock, What could we say here? Could we say, "You need to redesign/debug your handler"?

steveperry-53 replied:
@timstclair, Can  you comment on this.

chenopis commented on docs/concepts/workloads/pods/pod-lifecycle.md:
```diff
@@ -0,0 +1,281 @@
+   If the readiness probe fails, the endpoints controller removes the Pod's IP
+   address from the endpoints of all Services that match the Pod. The default
+   state of readiness before the initial delay is `Failure`. If a Container does
+   not provide a readiness probe, the dfault state is `Success`.
```
typo: dfault -> default

pwittrock commented:
Sorry DOSed.

@timstclair Does this seem like something you should look at?

steveperry-53 commented:
@timstclair, ping

timstclair commented:
Sorry, I've been production oncall for the past week. I can take a look tomorrow.

timstclair commented on docs/concepts/workloads/pods/pod-lifecycle.md:
```diff
@@ -0,0 +1,281 @@
+object, which has a `phase` field.
+
+Pod phases are consistent with the overall Kubernetes
+[API convention](https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md).
```
I think this can be removed for user-facing docs

steveperry-53 replied:
Done.

timstclair commented on docs/concepts/workloads/pods/pod-lifecycle.md:
```diff
@@ -0,0 +1,281 @@
+
+If any Container in a Pod specifies a readiness probe, the Pod reports a
+[PodCondition](/docs/resources-reference/v1.5/#podcondition-v1).
+A PodCondition has a `status` field that can be True, False, or Unknown.
```
This section isn't accurate. A pod has multiple conditions (currently Initialized, Ready, and PodScheduled, defined [here](https://github.com/kubernetes/kubernetes/blob/99ee9b4b45bf2d4760783bbf8bc86f7ef1581741/pkg/api/types.go#L1521)), and they are present whether or not any probes are specified.

steveperry-53 replied:
I took a stab at a rewrite.

timstclair commented on docs/concepts/workloads/pods/pod-lifecycle.md:
```diff
@@ -0,0 +1,281 @@
+## Container probes
+
+A [Probe](/docs/resources-reference/v1.5/#probe-v1) is a diagnostic
+performed periodically by the kubelet on a Container. To perform a diagnostic,
```
nit: why is `Container` capitalized?

timstclair replied:
Do we define kubelet anywhere? If so, it might be good to link to it here. If not, we should....

timstclair replied:
I just realized it's capitalized everywhere in this document. Is that a convention we follow everywhere?

steveperry-53 replied:
Going forward, our [style](https://kubernetes.io/docs/contribute/style-guide/#use-camel-case-for-api-objects) is to use the same case as the API object: Container, PodSpec, etc.

steveperry-53 replied:
Linked.

timstclair commented on docs/concepts/workloads/pods/pod-lifecycle.md:
```diff
@@ -0,0 +1,281 @@
+`restartPolicy` applies to all Containers in the Pod. `restartPolicy` only
+refers to restarts of the Containers by the kubelet on the same node. Failed
+Containers that are restarted by the kubelet are restarted with an exponential
+back-off delay. The delay is in multiples of sync-frequency (
```
I don't know what sync-frequency refers to, but the backoff is defined as 10 seconds. (i.e. 10s, 20s, 40s, etc.)

steveperry-53 replied:
Fixed.

timstclair commented on docs/concepts/workloads/pods/pod-lifecycle.md:
```diff
@@ -0,0 +1,281 @@
+[Pods document](/docs/user-guide/pods/#durability-of-pods-or-lack-thereof),
+once bound to a node, a Pod will never be rebound to another node. This means
+that some kind of controller is necessary for a Pod to survive node failure,
+even if just a single Pod at a time is desired.
```
This sentence seems out of place here. It looks redundant with the `Pod lifetime` section below?

steveperry-53 replied:
Fixed.

timstclair commented on docs/concepts/workloads/pods/pod-lifecycle.md:
```diff
@@ -0,0 +1,281 @@
+that some kind of controller is necessary for a Pod to survive node failure,
+even if just a single Pod at a time is desired.
+
+Three types of controllers are available:
```
I don't think the following belongs in the `Restart policy` section. Maybe the next section?

steveperry-53 replied:
Moved to next section.

timstclair commented on docs/concepts/workloads/pods/pod-lifecycle.md:
```diff
@@ -0,0 +1,281 @@
+  `restartPolicy` equal to OnFailure or Never.
+
+- Use a [ReplicationController](/docs/user-guide/replication-controller/) or
+  [Deployment](/docs/user-guide/deployments/)
```
Or ReplicaSet

steveperry-53 replied:
Added ReplicSet.

timstclair commented on docs/concepts/workloads/pods/pod-lifecycle.md:
```diff
@@ -0,0 +1,281 @@
+- Use a [DaemonSet](/docs/admin/daemons/) for Pods that need to run one per
+  machine, because they provide a machine-specific system service.
+
+All three types of controllers contain a PodTemplate, which is also in a Pod. It
```
nit: this sentence dosen't make sense. A PodTemplate is very similar to a pod, but the template isn't in the Pod (the templates metadata & spec are).

steveperry-53 replied:
Fixed, I think.

steveperry-53 commented:
@timstclair, Thanks for the review. I've addressed your comments inhttps://github.com/kubernetes/kubernetes.github.io/pull/2420/commits/f105893de535a29206a876be824999d11ed1ef9c.

timstclair commented:
Updates look good. Thanks!

/lgtm

chenopis commented:
@steveperry-53 If you rebase, that should fix the Travis CI issue.

maelvls commented on docs/concepts/workloads/pods/pod-lifecycle.md:
```diff
@@ -0,0 +1,282 @@
+of the PodCondition array has a `type` field and a `status` field. The `type`
+field is a string, with possible values PodScheduled, Ready, Initialized, and
+Unschedulable. The `status` field is a string, with possible values True, False,
+and Unknown.
```
Hi!
Sorry to bother 😁 I traced the addition of the 'Unschedulable' condition type to this commit.
But I can't find anywhere where the condition 'Unschedulable' is set (either in the kubelet or in the scheduler.go).

Is 'Unschedulable' just a reason of PodScheduled?
Or is it also a condition type?
